.dotnet/src/Utility/StreamingClientResult.cs

```diff
@@ -0,0 +1,160 @@
+using System;
+using System.ClientModel.Primitives;
+using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+using System.Runtime.CompilerServices;
+using System.Text.Json;
+using System.Threading;
+
+namespace OpenAI;
+
+/// <summary>
+/// Represents an operation response with streaming content that can be deserialized and enumerated while the response
+/// is still being received.
+/// </summary>
+/// <typeparam name="T"> The data type representative of distinct, streamable items. </typeparam>
+public class StreamingClientResult<T>
+    : IDisposable
+    , IAsyncEnumerable<T>
+{
+    private readonly PipelineResponse _baseResponse;
+    private readonly IAsyncEnumerable<T> _asyncEnumerable;
+    private bool _disposedValue;
+
+    /// <summary>
+    /// Gets the underlying <see cref="PipelineResponse"/> that contains headers and other response-wide information.
+    /// </summary>
+    /// <returns>
+    /// The <see cref="PipelineResponse"/> instance used in this <see cref="StreamingClientResult{T}"/>.
+    /// </returns>
+    public PipelineResponse GetRawResponse() => _baseResponse;
+
+    private StreamingClientResult() { }
+
+    private StreamingClientResult(
+        PipelineResponse response,
+        Func<Stream, IAsyncEnumerable<T>> asyncEnumerableProcessor)
+    {
+        _baseResponse = response;
+        _asyncEnumerable = asyncEnumerableProcessor.Invoke(_baseResponse.ContentStream);
+    }
+
+    /// <summary>
+    /// Creates a new instance of <see cref="StreamingClientResult{T}"/> that will yield items of the specified type
+    /// <typeparamref name="T"/> as they become available via server-sent event JSON data on the available
+    /// <see cref="PipelineResponse.ContentStream"/>. This overload supports deserializing multiple instaces of
+    /// <typeparamref name="T"/> per server-sent event using the provided multi-element deserialization delegate.
+    /// </summary>
+    /// <param name="response"> The base <see cref="PipelineResponse"/> for this result instance. </param>
+    /// <param name="multiElementJsonDeserializerFunc">
+    /// The delegate that will be used to extract a collection of elements from each incoming JSON data payload.
+    /// </param>
+    /// <param name="cancellationToken">
+    /// The optional cancellation token used to control the enumeration.
+    /// </param>
+    /// <returns> A new instance of <see cref="StreamingClientResult{T}"/>. </returns>
+    public static StreamingClientResult<T> Create(
+        PipelineResponse response,
+        Func<JsonElement, IEnumerable<T>> multiElementJsonDeserializerFunc,
```

Comment:
Why do we need this parameter? Why the method below (with the constraint to IJsonModel) is not enoughl? In addition, we would need a Create method overload that returns StreamingClientResult<BinaryData> for the protocol method case, i.e. when the caller want to get access to raw bytes of each event data so they can deserialize additional properties themselves.